File: server/src/test/java/de/ingrid/igeserver/db/LinkedDataTest.java
package de.ingrid.igeserver.db;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.orientechnologies.orient.core.db.ODatabaseSession;
import com.orientechnologies.orient.core.db.ODatabaseType;
import com.orientechnologies.orient.core.db.OrientDB;
import com.orientechnologies.orient.core.db.OrientDBConfig;
import com.orientechnologies.orient.core.metadata.schema.OClass;
import com.orientechnologies.orient.core.metadata.schema.OType;
import com.orientechnologies.orient.core.record.ORecord;
import com.orientechnologies.orient.core.record.impl.ODocument;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class LinkedDataTest {

    private static OrientDB db;

    private static OrientDBDatabase dbService = new OrientDBDatabase();
    private static ODatabaseSession testDB;

    @BeforeClass
    public static void before() {

        db = new OrientDB("memory:test", OrientDBConfig.defaultConfig());

        db.create("test", ODatabaseType.MEMORY);

        testDB = db.open("test", "admin", "admin");

        testDB.getMetadata().getSchema().createClass("Documents");
        testDB.getMetadata().getSchema().createClass("Addresses");
        testDB.commit();
    }

    @Before
    public void cleanup() {
        for (ODocument doc: testDB.browseClass("Documents")) {
            testDB.delete(doc);
        }
        for (ODocument address: testDB.browseClass("Addresses")) {
            testDB.delete(address);
        }
        testDB.commit();
    }

    @Test
    public void createLink() {

        ODocument address = new ODocument("Addresses");
        address.field("name", "André");
        address.field("city", "Frankfurt");
        ODocument dbAddress = address.save();

        ODocument doc = new ODocument("Documents");
        doc.field("title", "Dokument 1");
        doc.field("address", dbAddress);
        doc.save();

        testDB.commit();

        String json = testDB.browseClass("Documents").next().toJSON("fetchPlan:*:-1");
        assertEquals("{\"address\":{\"city\":\"Frankfurt\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}", json);

    }


    @Test
    public void updateLink() {
        ODocument address = new ODocument("Addresses");
        address.field("name", "André");
        address.field("city", "Frankfurt");
        ODocument dbAddress = address.save();

        ODocument doc = new ODocument("Documents");
        doc.field("title", "Dokument 1");
        doc.field("address", dbAddress);
        doc.save();

        testDB.commit();

        Optional<ORecord> query = testDB.query("SELECT * FROM Documents WHERE title='Dokument 1'").next().getRecord();

        assertTrue(query.isPresent());

        ODocument docFromDb = (ODocument) query.get();
        Map<String, Object> map = docFromDb.toMap();

        map.put("address.city", "Berlin");

        dbService.save(DBApi.DBClass.Documents.name(), null, map);

        //docFromDb.field("address.city", "Berlin");
        //docFromDb.save();

        testDB.commit();

        String json = testDB.browseClass("Documents").next().toJSON("fetchPlan:*:-1");
        assertEquals("{\"address\":{\"city\":\"Berlin\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}", json);

        assertEquals(1, testDB.countClass("Documents"));
        assertEquals(1, testDB.countClass("Addresses"));

    }

    @Test
    public void updateLinkMerge() {
        ODocument address = new ODocument("Addresses");
        address.field("name", "André");
        address.field("city", "Frankfurt");
        ODocument dbAddress = address.save();

        ODocument doc = new ODocument("Documents");
        doc.field("title", "Dokument 1");
        doc.field("address", dbAddress);
        doc.save();

        testDB.commit();

        Optional<ORecord> query = testDB.query("SELECT * FROM Documents WHERE title='Dokument 1'").next().getRecord();

        assertTrue(query.isPresent());

        ODocument docFromDb = (ODocument) query.get();
//        Map<String, Object> map = docFromDb.toMap();

//        map.put("address.city", "Berlin");

        docFromDb.field("address.city", "Berlin");
        docFromDb.save();

        testDB.commit();

        String json = testDB.browseClass("Documents").next().toJSON("fetchPlan:*:-1");
        assertEquals("{\"address\":{\"city\":\"Berlin\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}", json);

        assertEquals(1, testDB.countClass("Documents"));
        assertEquals(1, testDB.countClass("Addresses"));

    }

    @Test
    @Ignore
    public void updateLinkOnlyRID() {
        ODocument address = new ODocument("Addresses");
        address.field("name", "André");
        address.field("city", "Frankfurt");
        ODocument dbAddress = address.save();
        ODocument address2 = new ODocument("Addresses");
        address2.field("name", "Peter");
        address2.field("city", "München");
        ODocument dbAddress2 = address2.save();

        ODocument doc = new ODocument("Documents");
        doc.field("title", "Dokument 1");
        doc.field("description", "Beschreibung des Dokuments");
        doc.field("address", dbAddress);
        ODocument dbDoc = doc.save();

        testDB.commit();

        String json = dbDoc.toJSON("rid,version,fetchPlan:*:-1");
        Map<String, Object> map = dbDoc.toMap();


        ODocument updatedDoc = new ODocument();

        String modJson = "{\"@rid\":\"#17:0\",\"@version\":1,\"address\":{\"@rid\":\"#21:0\",\"@version\":1},\"title\":\"Dokument 1\"}";
        updatedDoc.fromJSON(modJson);

        // updatedDoc.save();
        ODocument dbMerged = dbDoc.merge(updatedDoc, true, true);
        testDB.commit();



        String jsonAfter = testDB.browseClass("Documents").next().toJSON("fetchPlan:*:-1");
        assertEquals("{\"address\":{\"city\":\"Berlin\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}", jsonAfter);

        assertEquals(1, testDB.countClass("Documents"));
        assertEquals(1, testDB.countClass("Addresses"));

    }

    @Test
    public void linktest() throws IOException {
        OClass addressClass = testDB.createClass("Address");
        OClass documentClass = testDB.createClass("Document");
        documentClass.createProperty("address", OType.LINK, addressClass);


        ODocument address = new ODocument("Address");
        address.field("city", "Frankfurt");
        address.field("country", "Germany");
        ODocument dbAddress1 = address.save();
        ODocument address2 = new ODocument("Address");
        address2.field("city", "Paris");
        address2.field("country", "France");
        ODocument dbAddress2 = address2.save();

        ODocument doc = new ODocument("Document");
        doc.field("title", "Name des Dokuments");
        doc.field("description", "Beschreibung");
        doc.field("address", dbAddress1.getIdentity());
        ODocument dbDoc1 = doc.save();

        testDB.commit();

        ODocument dbDoc = testDB.browseClass("Document").next();

        assertEquals("{\"@rid\":\"" + dbDoc1.getIdentity() + "\",\"address\":{\"@rid\":\"" + dbAddress1.getIdentity() + "\",\"country\":\"Germany\",\"city\":\"Frankfurt\"},\"description\":\"Beschreibung\",\"title\":\"Name des Dokuments\"}", dbDoc.toJSON("rid,fetchPlan:*:-1"));

//        dbDoc.field("address", dbAddress2.getIdentity());
        dbDoc.field("address", "34:0");
//        HashMap<Object, Object> map = new HashMap<>();
//        map.put("@rid", "#34:0");
//        dbDoc.field("address", map);

        assertEquals("{\"@rid\":\"" + dbDoc1.getIdentity() + "\",\"address\":{\"@rid\":\"" + dbAddress2.getIdentity() + "\",\"country\":\"France\",\"city\":\"Paris\"},\"description\":\"Beschreibung\",\"title\":\"Name des Dokuments\"}", dbDoc.toJSON("rid,fetchPlan:*:-1"));


        String inputJson = "{\"@rid\":\"" + dbDoc1.getIdentity() + "\",\"address\":{\"@rid\":\"" + dbAddress2.getIdentity() + "\"},\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}";

        ODocument docFromJson = new ODocument("Document");
        docFromJson.fromJSON(inputJson);
        assertEquals("{\"@rid\":\"" + dbDoc1.getIdentity() + "\",\"address\":{\"@rid\":\"" + dbAddress2.getIdentity() + "\",\"country\":\"France\",\"city\":\"Paris\"},\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}", docFromJson.toJSON("rid,fetchPlan:*:-1"));


    }

    @Test
    public void linkTestReal() throws IOException {

        // ========================================
        // DOCUMENT TYPE CLASS
        // ========================================

        // definition created from document type docType
        // {
        //   "geoservice": {
        //     "dbClass": "DOC_GEOSERVICE",
        //     "links": ["address"]
        //   },
        //   "address": {
        //     "dbClass": "MyAddress"
        //   }
        // }
        HashMap<String, Object> typeMapper = new HashMap<>();
        HashMap<String, Object> geoServiceTypeInfo = new HashMap<>();
        geoServiceTypeInfo.put("dbClass", "DOC_GEOSERVICE");
        geoServiceTypeInfo.put("links", new String[] {"address"});
        typeMapper.put("geoservice", geoServiceTypeInfo);
        HashMap<String, Object> addressTypeInfo = new HashMap<>();
        addressTypeInfo.put("dbClass", "MyAddress");
        typeMapper.put("address", addressTypeInfo);

        // creation of neccessary db classes with properties (links)
        for (String docType : typeMapper.keySet()) {
            HashMap profile = ((HashMap)typeMapper.get(docType));
            OClass documentClass = testDB.createClass((String) profile.get("dbClass"));
            String[] links = (String[]) profile.get("links");
            if (links != null) {
                for (String link : links) {
                    documentClass.createProperty(link, OType.LINK);
                }
            }
        }

        // ========================================
        // JUST SOME DEMO DATA
        // ========================================

        // set initial db data
        ODocument address = new ODocument("MyAddress");
        address.field("city", "Frankfurt");
        address.field("country", "Germany");
        ODocument dbAddress1 = address.save();

        ODocument address2 = new ODocument("MyAddress");
        address2.field("city", "Paris");
        address2.field("country", "France");
        ODocument dbAddress2 = address2.save();

        ODocument doc = new ODocument("DOC_GEOSERVICE");
        doc.field("title", "Name des Dokuments");
        doc.field("description", "Beschreibung");
        doc.field("address", dbAddress1.getIdentity());
        ODocument dbDoc1 = doc.save();

        // ========================================
        // CONTROLLER CLASS
        // ========================================

        // incoming test json from frontend
        //language=JSON
        String inputJson = "{\"@rid\":\"" + dbDoc1.getIdentity() + "\",\"_type\":\"geoservice\",\"address\":{\"@rid\":\"" + dbAddress2.getIdentity() + "\", \"city\":\"Paris\",\"country\":\"New France\"},\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}";

        // convert json to Node object
        ObjectNode node = (ObjectNode) new ObjectMapper().readTree(inputJson);

        // get document type from json node
        String docType = node.get("_type").asText();
        HashMap<String, Object> profileInfo = (HashMap<String, Object>) typeMapper.get(docType);

        // transform incoming document to replace nested documents with reference ID
        String[] links = (String[]) profileInfo.get("links");
        for (String link : links) {
            // set link reference into field directly
            node.put(link, node.get(link).get("@rid").textValue());
        }

        // create new db document object with defined class from document type
        String dbClass = (String) profileInfo.get("dbClass");

        // ========================================
        // DATABASE CLASS
        // ========================================
        ODocument docFromJson = new ODocument(dbClass);
        docFromJson.fromJSON(node.toString());

        assertEquals("{\"@rid\":\"" + dbDoc1.getIdentity() + "\",\"address\":{\"@rid\":\"" + dbAddress2.getIdentity() + "\",\"country\":\"France\",\"city\":\"Paris\"},\"_type\":\"geoservice\",\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}", docFromJson.toJSON("rid,fetchPlan:*:-1"));

        assertEquals(2, testDB.countClass("MyAddress"));
        assertEquals(1, testDB.countClass("DOC_GEOSERVICE"));

    }

}

File: server/src/test/java/de/ingrid/igeserver/persistence/orientdb/LinkedDataTest.kt
package de.ingrid.igeserver.persistence.orientdb

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode
import com.orientechnologies.orient.core.db.*
import com.orientechnologies.orient.core.metadata.schema.OType
import com.orientechnologies.orient.core.record.impl.ODocument
import de.ingrid.igeserver.persistence.orientdb.OrientDBDatabase
import de.ingrid.igeserver.persistence.orientdb.model.document.OAddressType
import de.ingrid.igeserver.persistence.orientdb.model.document.ODocumentType
import io.kotest.core.spec.style.FunSpec
import io.kotest.core.test.TestCase
import io.kotest.matchers.shouldBe
import java.util.*

class LinkedDataTest : FunSpec() {

    private val DOCUMENT_CLASS = ODocumentType().className
    private val ADDRESS_CLASS = OAddressType().className

    private val dbService = OrientDBDatabase()
    private lateinit var testDB: ODatabaseSession

    override fun beforeTest(testCase: TestCase) {
        val db = OrientDB("memory:test", OrientDBConfig.defaultConfig())
        db.create("test", ODatabaseType.MEMORY)
        testDB = db.open("test", "admin", "admin")
        testDB.metadata.schema.createClass(DOCUMENT_CLASS)
        testDB.metadata.schema.createClass(ADDRESS_CLASS)
        for (doc in testDB.browseClass(DOCUMENT_CLASS)) {
            testDB.delete(doc)
        }
        for (address in testDB.browseClass(ADDRESS_CLASS)) {
            testDB.delete(address)
        }
        testDB.commit()
    }

    init {
        dbService.entityTypes = listOf(ODocumentType(), OAddressType())

        test("create link as value")
        {
            // create document with linked address
            val address = ODocument(ADDRESS_CLASS)
            address.field("name", "André")
            address.field("city", "Frankfurt")
            val dbAddress = address.save()
            val doc = ODocument(DOCUMENT_CLASS)
            doc.field("title", "Dokument 1")
            doc.field("address", dbAddress)
            doc.save()
            testDB.commit()

            val json = testDB.browseClass(DOCUMENT_CLASS).next().toJSON("fetchPlan:*:-1")
            json shouldBe "{\"address\":{\"city\":\"Frankfurt\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}"
        }

        test("update content of link as value")
        {
            // create document with linked address
            val address = ODocument(ADDRESS_CLASS)
            address.field("name", "André")
            address.field("city", "Frankfurt")
            val dbAddress = address.save()
            val doc = ODocument(DOCUMENT_CLASS)
            doc.field("title", "Dokument 1")
            doc.field("address", dbAddress)
            doc.save()
            testDB.commit()

            // update linked address
            val query = testDB.query("SELECT * FROM $DOCUMENT_CLASS WHERE title='Dokument 1'").next().record
            query.isPresent shouldBe true
            val docFromDb = query.get() as ODocument

            // TODO using DBApi does not update the linked content
            //val map = docFromDb.toMap()
            //map["address.city"] = "Berlin"
            //dbService.save(DocumentType::class, null, "{\"address\": { \"city\": \"Berlin\"}}", null)

            docFromDb.field("address.city", "Berlin");
            docFromDb.save();
            testDB.commit()

            val json = testDB.browseClass(DOCUMENT_CLASS).next().toJSON("fetchPlan:*:-1")
            json shouldBe "{\"address\":{\"city\":\"Berlin\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}"
            testDB.countClass(DOCUMENT_CLASS) shouldBe 1
            testDB.countClass(ADDRESS_CLASS) shouldBe 1
        }

        test("update content of link as value (duplicate?)")
        {
            // create document with linked address
            val address = ODocument(ADDRESS_CLASS)
            address.field("name", "André")
            address.field("city", "Frankfurt")
            val dbAddress = address.save()
            val doc = ODocument(DOCUMENT_CLASS)
            doc.field("title", "Dokument 1")
            doc.field("address", dbAddress)
            doc.save()
            testDB.commit()

            // update linked address
            val query = testDB.query("SELECT * FROM $DOCUMENT_CLASS WHERE title='Dokument 1'").next().record
            query.isPresent shouldBe true
            val docFromDb = query.get() as ODocument

            // TODO what is the difference to preceding test?
            //val map = docFromDb.toMap();
            //map.put("address.city", "Berlin");
            docFromDb.field("address.city", "Berlin")
            docFromDb.save()
            testDB.commit()

            val json = testDB.browseClass(DOCUMENT_CLASS).next().toJSON("fetchPlan:*:-1")
            json shouldBe "{\"address\":{\"city\":\"Berlin\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}"
            testDB.countClass(DOCUMENT_CLASS) shouldBe 1
            testDB.countClass(ADDRESS_CLASS) shouldBe 1
        }

        test("update content of link as value by changing @rid?").config(enabled = false)
        {
            // TODO fix test, what is the intention of the test?

            // create document with linked address and extra address
            val address = ODocument(ADDRESS_CLASS)
            address.field("name", "André")
            address.field("city", "Frankfurt")
            val dbAddress = address.save()
            val address2 = ODocument(ADDRESS_CLASS)
            address2.field("name", "Peter")
            address2.field("city", "München")
            val dbAddress2 = address2.save()
            val doc = ODocument(DOCUMENT_CLASS)
            doc.field("title", "Dokument 1")
            doc.field("description", "Beschreibung des Dokuments")
            doc.field("address", dbAddress)
            val dbDoc = doc.save()
            testDB.commit()

            // link second address with document
            val json = dbDoc.toJSON("rid,version,fetchPlan:*:-1")
            val map = dbDoc.toMap()
            val updatedDoc = ODocument()
            val modJson = "{\"@rid\":\"#17:0\",\"@version\":1,\"address\":{\"@rid\":\"#21:0\",\"@version\":1},\"title\":\"Dokument 1\"}"
            updatedDoc.fromJSON(modJson)
            // updatedDoc.save();
            val dbMergedDOCUMENT_CLASS = dbDoc.merge(updatedDoc, true, true)
            testDB.commit()

            val jsonAfter = testDB.browseClass(DOCUMENT_CLASS).next().toJSON("fetchPlan:*:-1")
            jsonAfter shouldBe "{\"address\":{\"city\":\"Berlin\",\"name\":\"Andr\\u00e9\"},\"title\":\"Dokument 1\"}"
            testDB.countClass(DOCUMENT_CLASS) shouldBe 1
            testDB.countClass(ADDRESS_CLASS) shouldBe 1
        }

        test("create link as OType.LINK")
        {
            // setup relation
            val addressClass = testDB.getClass(ADDRESS_CLASS)
            val documentClass = testDB.getClass(DOCUMENT_CLASS)
            documentClass.createProperty("address", OType.LINK, addressClass)

            // create document with linked address and extra address
            val address = ODocument(ADDRESS_CLASS)
            address.field("city", "Frankfurt")
            address.field("country", "Germany")
            val dbAddress1 = address.save()
            val address2 = ODocument(ADDRESS_CLASS)
            address2.field("city", "Paris")
            address2.field("country", "France")
            val dbAddress2 = address2.save()
            val doc = ODocument(DOCUMENT_CLASS)
            doc.field("title", "Name des Dokuments")
            doc.field("description", "Beschreibung")
            doc.field("address", dbAddress1.identity)
            val dbDoc1 = doc.save()
            testDB.commit()

            val dbDoc = testDB.browseClass(DOCUMENT_CLASS).next()
            dbDoc.toJSON("rid,fetchPlan:*:-1") shouldBe "{\"@rid\":\"${dbDoc1.identity}\",\"address\":{\"@rid\":\"${dbAddress1.identity}\",\"country\":\"Germany\",\"city\":\"Frankfurt\"},\"description\":\"Beschreibung\",\"title\":\"Name des Dokuments\"}"

            // link to second address (set field)
            //dbDoc.field("address", dbAddress2.getIdentity());
            dbDoc.field("address", "34:0")
            //var map = mutableMapOf<String, String>();
            //map["@rid"] = "#34:0";
            //dbDoc.field("address", map);
            dbDoc.toJSON("rid,fetchPlan:*:-1") shouldBe "{\"@rid\":\"${dbDoc1.identity}\",\"address\":{\"@rid\":\"${dbAddress2.identity}\",\"country\":\"France\",\"city\":\"Paris\"},\"description\":\"Beschreibung\",\"title\":\"Name des Dokuments\"}"

            // link to second address (from json)
            val inputJson = "{\"@rid\":\"${dbDoc1.identity}\",\"address\":{\"@rid\":\"${dbAddress2.identity}\"},\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}"
            val docFromJson = ODocument(DOCUMENT_CLASS)
            docFromJson.fromJSON(inputJson)
            docFromJson.toJSON("rid,fetchPlan:*:-1") shouldBe "{\"@rid\":\"${dbDoc1.identity}\",\"address\":{\"@rid\":\"${dbAddress2.identity}\",\"country\":\"France\",\"city\":\"Paris\"},\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}"
        }

        test("tbd real")
        {
            // ========================================
            // DOCUMENT TYPE CLASS
            // ========================================

            // definition created from document type docType
            // {
            //   "geoservice": {
            //     "dbClass": "DOC_GEOSERVICE",
            //     "links": ["address"]
            //   },
            //   "address": {
            //     "dbClass": "MyAddress"
            //   }
            // }
            val typeMapper = HashMap<String, Any>()
            val geoServiceTypeInfo = HashMap<String, Any>()
            geoServiceTypeInfo["dbClass"] = "DOC_GEOSERVICE"
            geoServiceTypeInfo["links"] = arrayOf("address")
            typeMapper["geoservice"] = geoServiceTypeInfo
            val addressTypeInfo = HashMap<String, Any>()
            addressTypeInfo["dbClass"] = "MyAddress"
            typeMapper["address"] = addressTypeInfo

            // creation of necessary db classes with properties (links)
            for (docType in typeMapper.keys) {
                val profile = typeMapper[docType] as HashMap<*, *>?
                val documentClass = testDB.createClass(profile!!["dbClass"] as String?)
                val links = profile["links"] as Array<String>?
                if (links != null) {
                    for (link in links) {
                        documentClass.createProperty(link, OType.LINK)
                    }
                }
            }

            // ========================================
            // JUST SOME DEMO DATA
            // ========================================

            // set initial db data
            val address = ODocument("MyAddress")
            address.field("city", "Frankfurt")
            address.field("country", "Germany")
            val dbAddress1 = address.save()
            val address2 = ODocument("MyAddress")
            address2.field("city", "Paris")
            address2.field("country", "France")
            val dbAddress2 = address2.save()
            val doc = ODocument("DOC_GEOSERVICE")
            doc.field("title", "Name des Dokuments")
            doc.field("description", "Beschreibung")
            doc.field("address", dbAddress1.identity)
            val dbDoc1 = doc.save()

            // ========================================
            // CONTROLLER CLASS
            // ========================================

            // incoming test json from frontend
            //language=JSON
            val inputJson = "{\"@rid\":\"${dbDoc1.identity}\",\"_type\":\"geoservice\",\"address\":{\"@rid\":\"${dbAddress2.identity}\", \"city\":\"Paris\",\"country\":\"New France\"},\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}"

            // convert json to Node object
            val node = ObjectMapper().readTree(inputJson) as ObjectNode

            // get document type from json node
            val docType = node["_type"].asText()
            val profileInfo = typeMapper[docType] as HashMap<String, Any>?

            // transform incoming document to replace nested documents with reference ID
            val links = profileInfo!!["links"] as Array<String>?
            for (link in links!!) {
                // set link reference into field directly
                node.put(link, node[link]["@rid"].textValue())
            }

            // create new db document object with defined class from document type
            val dbClass = profileInfo["dbClass"] as String?

            // ========================================
            // DATABASE CLASS
            // ========================================
            val docFromJson = ODocument(dbClass)
            docFromJson.fromJSON(node.toString())
            docFromJson.toJSON("rid,fetchPlan:*:-1") shouldBe "{\"@rid\":\"${dbDoc1.identity}\",\"address\":{\"@rid\":\"${dbAddress2.identity}\",\"country\":\"France\",\"city\":\"Paris\"},\"_type\":\"geoservice\",\"description\":\"Beschreibung\",\"title\":\"Neuer Name des Dokuments\"}"
            testDB.countClass("MyAddress") shouldBe 2
            testDB.countClass("DOC_GEOSERVICE") shouldBe 1
        }
    }
}